torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -377,14 +377,16 @@ void ProcessGroupAgent::enqueueRecv(RecvWork work) {
         if (message.isRequest()) {
           auto futureResponse = cb_->operator()(message);
           if (futureResponse->completed()) {
-            send(work.from_, std::move(*futureResponse).moveMessage());
+            send(work.from_, std::move(*futureResponse).moveValue());
```

Comment:
One question that I have about the various moveValue() cases (here, 9 lines down, and parallel two cases in ThriftRpcAgent) is that we probably need to deal with the setError case.

In all these cases, it is quite possible that futureResponse was fulfilled with setError(), rather than with markCompleted(). Right now, it does seem like there's even a hasError() accessor in Future<T>?

I imagine all we need is a short helper like the following:
```
  Message packForSending(shared_ptr<FutureMessage> future) {
      if (future->hasError()) {
        return Message m(..., MessageType::EXCEPTION);
      } else {
        return std::move(*futureResponse).moveValue();
      }
  }
```

Then it something like 
```
      if (futureResponse->completed()) { 
         send(work.from_, packForSending(futureResponse));
      } 

```    
etc in the 4 rpc_agent places.

Reply:
@jjlilley that is great catch! I think what we need for futureMessage case is to always markCompleted(Message) then because our system may need to do some special handling based message exception type (should be rare case because we rely on error_ inside future to do error handling now).

Reply:
Oh, make sense - so we can choose to avoid the exception path for Future<Message>, but for generic Future<T> still have this feature.

Reply:
It still might be nice to have the equivalent of a hasException() accessor like folly::Future

Reply:
sounds good! let me add the hasError() accessor

Reply:
actually we can not always markCompleted after setError, because setError is kind of markCompleted as well. 

but all the error handling are in future now, we do not need to send message with Exception type to client any more actually. I checked there is no usage in client side in our system right now.

Reply:
So, if tests pass right now as-is, I don't see this as a blocking issue for this change.

That said, the first time somebody in request_callback_impl chooses to return a setError(), this will fail. Also note that this is server-side processing, not client side.

So, if we skip here, I will write up a followup fix in this regard

Reply:
@jjlilley I'm just trying to understand why it will fail if someone setError() in requestCallback?  if someone setError(), the moveValue() will just move dummy Message(), right?